docs/content/sips/010-key-value.md

```diff
@@ -0,0 +1,60 @@
+title = "SIP 010 - Key-Value Storage"
+template = "main"
+date = "2023-01-23T01:01:01Z"
+---
+
+Summary: Provide a generic interface for access to various key-value storage systems.
+
+Owner(s): joel.dice@fermyon.com
+
+Created: Jan 23, 2023
+
+## Background
+
+Spin should have native support for using a variety of key-value (KV) datastores.
+
+Currently, users can use Redis for persistent key-value storage for Spin applications. This [documentation](https://developer.fermyon.com/cloud/data-redis.md) walks through how to use the existing Spin Redis SDK which leverages the `[outbound-redis.wit` interface](https://github.com/fermyon/spin/blob/main/wit/ephemeral/outbound-redis.wit). In contrast to the Redis interface, we are proposing a more general-purpose interface which can be used with a variety of implementations.
+
+## Proposal
+
+In order to support key-value stores, the following need to be added to Spin:
+
+- A `WIT` file that defines the key-value interface
+- SDK implementations for various programming languages
+- A default local key-value store
+
+Although not in scope for this proposal, we'll also want to expand the runtime configuration code added in [this PR](https://github.com/fermyon/spin/pull/798) to support configuring various key-value stores, including the default one.
+
+### Key-Value Interface (`.wit`)
+
+Spin should leverage the WebAssembly WASI subgroup’s work to define a universal key-value `WIT` interface. That work is taking place in [the `wasi-keyvalue` proposal](https://github.com/WebAssembly/wasi-keyvalue). However, the proposal was made assuming that `WIT` star imports are available, which is not yet the case. Therefore, we're modeling stores as pseudo-[resource handles](https://github.com/WebAssembly/component-model/blob/main/design/mvp/WIT.md#item-resource) which may be created using an `open` function and disposed using a `close` function.  Each operation on the store is a function which accepts a handle as its first parameter.
+
+```
+type store = u32
+
+variant error {
+  store-table-full,
+  no-such-store,
```

Comment:
Does this imply stores have to be created?  If so how does that happen?

Reply:
The current thinking is that only the default store (named "") is available default.  Other stores would need to be specified via runtime configuration.  If an app tries to open a store that is not the default and has not been configured, it will get `no-such-store`.

Reply:
I'll add doc comments to the WIT file to describe each function and the errors it can raise.